tests/validator/validator.cpp

```diff
@@ -698,70 +1448,349 @@ TEST(Validator, parseAndValidateInvalidUnitErrors) {
     v4->setUnits("dimensionless");
+	v5->setUnits("dimensionless");
+	v6->setUnits("dimensionless");
+    v7->setUnits("dimensionless");
 
     comp1->addVariable(v1_1);
     comp1->addVariable(v1_2);
     comp2->addVariable(v2);
     comp3->addVariable(v3);
     comp4->addVariable(v4);
+	comp5->addVariable(v5);
+	comp6->addVariable(v6);
+    comp7->addVariable(v7);
     m->addComponent(comp1);
     m->addComponent(comp2);
     m->addComponent(comp3);
     m->addComponent(comp4);
+	m->addComponent(comp5); 
+	m->addComponent(comp6);
+    m->addComponent(comp7);
 
     // Valid connections.
-    libcellml::Variable::addEquivalence(v1_1, v2);
+    libcellml::Variable::addEquivalence(v1_1, v2);  
     libcellml::Variable::addEquivalence(v1_2, v2);
-    libcellml::Variable::addEquivalence(v1_1, v3);
+    // Removed the line below to prevent the cycles from appearing here.  Cycles are tested separately in validateNoCycles
+    // libcellml::Variable::addEquivalence(v1_1, v3);  // !! creates cycle with v1_1, v2, v3
     libcellml::Variable::addEquivalence(v1_1, v4);
-    libcellml::Variable::addEquivalence(v2, v3);
-    libcellml::Variable::addEquivalence(v1_1, v3);
-    // Make v4 a variable without a parent component.
+    libcellml::Variable::addEquivalence(v2, v3);    
+	libcellml::Variable::addEquivalence(v5, v6);	// valid here but duplicated below
+
+	// Not valid connections
+	libcellml::Variable::addEquivalence(v6, v5);    // duplicated above, does not overwrite, skips silently
+    
+	// Make v4 a variable without a parent component.
     comp4->removeVariable(v4);
     // Remove all connections on v1_2, leaving dangling reciprocal connections.
     v1_2->removeAllEquivalences();
 
     v.validateModel(m);
+
     EXPECT_EQ(expectedErrors.size(), v.errorCount());
     for (size_t i = 0; i < v.errorCount(); ++i) {
-        EXPECT_EQ(expectedErrors.at(i), v.getError(i)->getDescription());
+       EXPECT_EQ(expectedErrors.at(i), v.getError(i)->getDescription());
+    }
+}
+
+TEST(Validator, validateConnectionComponent1NotEqualComponent2) {
+
+    std::vector<std::string> expectedErrors = {
+        "Variable 'doppelganger' has an equivalent variable 'doppelganger' equal to itself. ",
+    };
+    libcellml::Validator validator;
+    libcellml::ModelPtr m = std::make_shared<libcellml::Model>();
+    libcellml::ComponentPtr c = std::make_shared<libcellml::Component>();
+    libcellml::VariablePtr v = std::make_shared<libcellml::Variable>();
+    m->setName("modelName");
+    v->setName("doppelganger");
+    v->setUnits("dimensionless");
+    c->setName("bucket");
+    c->addVariable(v);
+    m->addComponent(c);
+    libcellml::Variable::addEquivalence(v, v);          
+    validator.validateModel(m);
+
+    EXPECT_EQ(1u, validator.errorCount());
+    for (size_t i = 0; i < validator.errorCount(); ++i) {
+        EXPECT_EQ(validator.getError(i)->getDescription(), expectedErrors[i]);
+    }
+}
+
+TEST(Validator, validateNoCycles) {
```

Comment:
I find this test is a bit more than required.  I like to see tests that produce the simplest scenario possible to test the desired result.

Reply:
Haha, agreed, but I needed to make sure that the validator could actually work in a complicated situation.  Will make another simpler one too :)